docs/lang/control-and-nondeterminism.md

```diff
@@ -0,0 +1,509 @@
+
+Why do we call it a device? We cannot call it a function, as functions are
+deterministic by definition. For the same reason, it is not a TLA+
+operator. In logic, we would say that `GUESS` is simply a binary relation on
+sets and their elements, which would not be different from the membership
+relation `\in`.
+
+Why do we need `GUESS S` and cannot use `CHOOSE x \in S: TRUE` instead?
+Actually, `CHOOSE x \in S: TRUE` is *deterministic*. It is guaranteed to return
+the same value, when it is called on two equals sets: if `S = T`, then
+`(CHOOSE x \in S: TRUE) = (CHOOSE x \in T: TRUE)`. Our `GUESS S` does not have
+this guarantee. It is free to return an arbitrary element of `S` each time
+we call it.
+
+How to implement `GUESS S`? There is no general answer to this question.
+However, we know of multiple sources of non-determinism in computer science. So
+we can think of `GUESS S` as being one of the following implementations:
+
+ 1. `GUESS S` can be a remote procedure call in a distributed system.  Unless,
+ we have centralized control over the distributed system, the returned value of
+ RPC may be non-deterministic.
+
+ 1. `GUESS S` can be simply the user input. In this case, the user resolves
+ non-determinism.
+
+ 1. `GUESS S` can be controlled by an adversary, who is trying to break the
+ system.
+
+ 1. `GUESS S` can pick an element by calling a random number generator.
+ However, note that RNG is a very special way of resolving non-determinism: It
+ assumes probabilistic distribution of elements (usually, it is close to the
+ [uniform
+ distribution](https://en.wikipedia.org/wiki/Discrete_uniform_distribution)).
+ Thus, the probability of producing an unfair choice of elements with RNG will
+ be approaching 0.
+
+
+As you see, there are multiple sources of non-determinism. With `GUESS S` we can
+model all of them. As TLA+ does not introduce special primitives for different
+kinds of non-determinism, neither do we fix any implementation of `GUESS S`.
+
+**Halting.** Note that `GUESS {}` halts the evaluation. What does it mean? The
+evaluation cannot continue in the current partial state. It does not imply that
+we have found a deadlock in our TLA+ specification. It simply means that we
+made wrong choices on the way. If we like to enumerate all possible state
+successors, like TLC does, we have to backtrack. In general, the course of
+action depends on the program analysis that you implement. For instance,
+a random simulator could simply backtrack and randomly choose another value.
+
+<a name="nondetExists"></a>
+### Non-determinism in `\E x \in S: P`
+
+We only have to consider the following case: `\E x \in S: P` is evaluated in a
+partial state `s`, and there is a primed state variable `y'` that satisfies two
+conditions:
+
+ 1. The predicate `P` refers to `y'`, that is, `P` has to assign a value to `y'`.
+ 2. The value of `y'` is not defined yet, that is, `s.y' = Null`.
```

Comment:
Continuation of the above, don't treat `s'` as if it's a variable and part of the same state as `s.y` , it's a different state's value `s_2.y`.

Reply:
That should be better now with the bindings